types/decimal.go

```diff
@@ -0,0 +1,403 @@
+	precisionMultipliers []*big.Int
+	zeroInt              = big.NewInt(0)
+	tenInt               = big.NewInt(10)
+)
+
+// Set precision multipliers
+func init() {
+	precisionMultipliers = make([]*big.Int, Precision+1)
+	for i := 0; i <= Precision; i++ {
+		precisionMultipliers[i] = calcPrecisionMultiplier(int64(i))
+	}
+}
+
+func precisionInt() *big.Int {
+	return new(big.Int).Set(precisionReuse)
+}
+
+// nolint - common values
+func ZeroDec() Dec { return Dec{new(big.Int).Set(zeroInt)} }
+func OneDec() Dec  { return Dec{precisionInt()} }
+
+// calculate the precision multiplier
+func calcPrecisionMultiplier(prec int64) *big.Int {
+	if prec > Precision {
+		panic(fmt.Sprintf("too much precision, maximum %v, provided %v", Precision, prec))
+	}
+	zerosToAdd := Precision - prec
+	multiplier := new(big.Int).Exp(tenInt, big.NewInt(zerosToAdd), nil)
+	return multiplier
+}
+
+// get the precision multiplier. Do not mutate result.
+func precisionMultiplier(prec int64) *big.Int {
+	if prec > Precision {
+		panic(fmt.Sprintf("too much precision, maximum %v, provided %v", Precision, prec))
+	}
+	return precisionMultipliers[prec]
+}
+
+//______________________________________________________________________________________________
+
+// create a new Dec from integer assuming whole number
+func NewDec(i int64) Dec {
+	return NewDecWithPrec(i, 0)
+}
+
+// create a new Dec from integer with decimal place at prec
+// CONTRACT: prec <= Precision
+func NewDecWithPrec(i, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(big.NewInt(i), precisionMultiplier(prec)),
+	}
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec <= Precision
+func NewDecFromBigInt(i *big.Int) Dec {
+	return NewDecFromBigIntWithPrec(i, 0)
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec <= Precision
+func NewDecFromBigIntWithPrec(i *big.Int, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(i, precisionMultiplier(prec)),
+	}
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec <= Precision
+func NewDecFromInt(i Int) Dec {
+	return NewDecFromIntWithPrec(i, 0)
+}
+
+// create a new Dec from big integer with decimal place at prec
+// CONTRACT: prec <= Precision
+func NewDecFromIntWithPrec(i Int, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(i.BigInt(), precisionMultiplier(prec)),
+	}
+}
+
+// create a decimal from an input decimal string.
+// valid must come in the form:
+//   (-) whole integers (.) decimal integers
+// examples of acceptable input include:
+//   -123.456
+//   456.7890
+//   345
+//   -456789
+//
+// NOTE an error will return if more decimal places
+// are provided in the string than the constant Precision
+func NewDecFromStr(str string) (d Dec, err Error) {
+	if len(str) == 0 {
+		return d, ErrUnknownRequest("decimal string is empty")
+	}
+
+	// first extract any negative symbol
+	neg := false
+	if string(str[0]) == "-" {
```

Comment:
Can we just compare the character directly?

Reply:
addressed